File: src/main/java/it/giaquinto/springberry/model/user/UserFactory.java
package it.giaquinto.springberry.model.user;

import it.giaquinto.springberry.model.security.SecurityLevel;
import it.giaquinto.springberry.model.user.implementation.*;

public abstract class UserFactory {
    public static BaseUser fromBaseUser(final BaseUser baseUser, final SecurityLevel level) {
        return switch(level) {
            case NONE -> null;
            case GUEST -> new GuestUser(baseUser.getName(), baseUser.getSurname(), baseUser.getCreated());
            case NORMAL -> new NormalUser(baseUser.getName(), baseUser.getSurname(), baseUser.getCreated());
            case POWER -> new PowerUser(baseUser.getName(), baseUser.getSurname(), baseUser.getCreated());
            case ADMIN -> new AdminUser(baseUser.getName(), baseUser.getSurname(), baseUser.getCreated());
        };
    }
}

File: src/main/java/it/giaquinto/springberry/model/user/UserFactory.kt
package it.giaquinto.springberry.model.user

import it.giaquinto.springberry.model.security.SecurityLevel
import it.giaquinto.springberry.model.user.implementation.*

object UserFactory {
    fun fromBaseUser(baseUser: BaseUser, level: SecurityLevel?): BaseUser {
        return when (level) {
            SecurityLevel.NONE -> null
            SecurityLevel.GUEST -> GuestUser(baseUser.name, baseUser.surname, baseUser.created)
            SecurityLevel.NORMAL -> NormalUser(baseUser.name, baseUser.surname, baseUser.created)
            SecurityLevel.POWER -> PowerUser(baseUser.name, baseUser.surname, baseUser.created)
            SecurityLevel.ADMIN -> AdminUser(baseUser.name, baseUser.surname, baseUser.created)
        }
    }
}